### ISSUE TITLE
Implement xor for int

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#2049]
I think this is part of the logical protocol. I guess we just haven't gotten around to implementing that, but at least doing xor would be very helpful! I think we had a use case where we converted via i64 just to be able to xor. @sydow is this something you want to look at? :)

**Comment by sydow** [comment]
My strong instinct is to say no. But there may be arguments that make me change my mind, so I would be interested in hearing more about the use case Kristian mentions. But the reason that int does not implement Logical is certainly not that we haven’t gotten around to doing it; it is intentional. Here is how I see things. 

The logical bit operations are useful in many situations, and it is natural and useful to extend them to bit arrays, as this is commonly supported by hardware and can be very efficient in both space and time. Thus, in many languages these operations can act on (bounded) integers of various sizes, with the obvious semantics that operations are done bitwise. This also means that these operations do not have a widely useful semantics for integer variables with two’s complement representation. So, typically, an integer variable is either used as a bit array (involved in bit operations only) or as an integer (involved in arithmetics). Special situations that extend this are often cited (e.g., on some architectures xor-ing a register with itself is the fastest way to set it to -1), but I think these are exceptions. 

Now to arbitrary size integers. First, we have a conceptual problem: What should  e.g. (-4) xor (-6) evaluate to in type int? The common take is to think of int’s as represented in two’s complement notation "with an infinite number of sign bits". Then everything becomes well defined. But bigints are typically not implemented in two’s complement, but have a sign/magnitude representation, so we lose both efficiency *and* the straightforward bit semantics. Why would we want this?

There may be a case for having bit operations for bignats (i.e., only positive numbers) but we do not have that type. Crypto also needs bit operations (in particular xor) on large data, but there one wants the true bit-by-bit operations, not the fake two’s complement version described above.

I leave it at this and wait for your comments.

**Comment by plajjan** [comment]
Well, given our current use of ints in a lot of places (sort of my fault) and type inference, we usually end up with numbers as `int` and then when you try to xor, you get an exception. It works in Python, so it just seemed like something we missed. It's also a bit ugly that it's a runtime error and not a type error.

I agree with negative integer problem in that it's hard to imagine what the result should be. I have actually attempted to implement xor for ints myself, got stuck on negative integers and then opened this issue because I just couldn't wrap my head around what would make sense (and I learnt about two's complement through that little journey).

We only loose efficiency on negative numbers, right? For positive numbers we just need to match limb length and do normal xor, no?

Is it so bad to do the same as Python or other languages that support bigints?

**Comment by sydow** [comment]
First of all, it is certainly not your fault that we overuse int for the moment. And we will hopefully soon fix that.

But I thought that variables that you use for xor-ing are typically not used for arithmetic, so you could declare such variables to be e.g. u64 without any problems.

I could also consider changing my mind, if you disagree. After all, to implement xor for int is not so complicated. As you say, for positive operands you just do word-wise xor-ing. If both operands are negative you use x ^ y = (~x) ^ (~y); if they have different signs you use a similar formula. And and or are actually a little bit more complicated, but not much.

**Comment by plajjan** [comment]
Yeah, I sort of agree that it seems reasonable that variables would be used either for arithmetic or bit-banging xor / and or whatnot, but I dunno, I haven't written enough code to say what happens in real world. However, it all just works rather conveniently in Python and given that it doesn't seem *that* hard to implement for our ints, I think we should.

I think it all boils down to the same arguments in the end, just like for any other arithmetic operations, there might be reasons to pick a different type, like u64, than whatever default the type inference will pick. We should have some convenient approach out of the box and the user might have to explicitly state a type to make a more informed choice. That doesn't stop us from supporting Logical for int.
